### ISSUE TITLE
Inference of clock model parameters for continuous data fails with MCMC flatlines, suspect interaction of FBD tree model and dnPhyloBrownianREML()

### ISSUE THREAD

**Issue by ixchelgzlzr** [revbayes/revbayes#565]
Hello RevBayes developers! 

I am unsure if the issue I am describing below is a bug or an user error (most likely in the Brownian model component), so I thank you so much in advance for any help you can provide!

I am setting up a Total Evidence Dating analysis on a dataset that contains molecular, morphological categorical, and morphological continuous data. 

**Model description:**
I am using an FBD tree model, standard GTR + G and Mk +G models of evolution for the molecular and categorical morphological data, and UCLN clock model for the molecular data. The clock of the categorical morphological data is linked to the molecular clock rates. For the continuous data, I am using a Brownian model, and trying to set up a global rate for the continuous data clock model. As an additional note, I have set up two topological constraints.

**Issue:**
The MCMCs flatline for all the parameters of the continuous characters evolution and clock models., e.g.:

![Screenshot 2024-10-23 at 12 45 09 PM](https://github.com/user-attachments/assets/d3947b38-8a1a-4ab3-b2ce-ef3003bddc31)

I have troubleshooted many options and here is what I have so far:

1) When I run an unrooted/no-time-divergence model, the parameters of the Brownian model are well sampled. ---> the problem seem to be specifically with the clock model for continuous data.
2) I turned off the moves on sampled ancestors and I still get the flatlines, so I don't think the issue is related to having sampled ancestors. 
3) When I don't include the continuous data, the MCMCs mix well ---> Another evidence is the continuous part of the model that isn't working. 
4) The continuous dataset has missing data (i.e. fossils), so I fillied in that missing data, and still got the flatlines --->  missing data might not be the issue. 

In summary, I suspect an interaction between the FBD tree prior and the molecular clock for the continuous data.  Alternatively, I might be setting this model wrong...

**To Reproduce**
I have set up a reduced example of my analysis here:
` https://github.com/ixchelgzlzr/bug_report_TED_solanaceae.git`

There are instructions on how to run the different versions of the analyses in the `Readme` file. 

I have also included some examples of output in `/output_examples`. Please notice I have set up a seed() that replicates the screenshot I pasted above.

**Expected behavior**
Mixing MCMCs for the continuous data parameters.

**Computer info**
I'm using RevBayes version (1.2.5-preview)
Build from remotes/origin/development (rapture-3049-g3bb16a)
in MacBook M1, with Sonoma 14.5

Thank you so much again!

**Comment by bjoelle** [comment]
Some observations about this issue
 * the tree is also getting stuck during the interval
 * removing the non-continuous data also removes the problem so I don't think it's specific to continuous data
 * the likelihoods etc appear to be calculated properly (also evidenced by the fact that it does get unstuck eventually), the moves are just getting rejected due to low posterior ratio

So in summary it doesn't appear to be a bug but rather the inference getting stuck in a local optima and then unable to move, which is a risk with very complex models where many things are interlinked

One thing that may help would be to adjust the priors or add constraints to stop it from getting in that position in the first place, if possible

Another thing that I observed is that running with all data or without the continuous data gives me a tree length of around 4000 whereas running just the continuous data gives me a tree length of 2000 (with around the same root age for all cases) - so it seems that the continuous part of the setup is driving towards a very different tree shape than the rest, which may be something to investigate

**Comment by bredelings** [comment]
I haven't managed to reproduce the trace above using `rb headers/MCMC/model_TED_w_backbone_const090624.Rev`.  However, I see that lots of parameters have shorter flat spots of around 5-20 iterations.

I decided to check for problems with the likelihood calculation by defining DEBUG_MCMC.  But I haven't found any problems yet, as @bjoelle predicted.  I think you could add this flag by running `export CXXFLAGS=-DDEBUG_MCMC` before you run cmake, if you want to check more.

One thing you can try is to replace some of the mvScale and mvSlide with mvSlice.  When the MCMC wants very small steps in one part of the space and large moves in other parts of the space, mvSlice works better.  However
- it won't work on node heights or branch lengths, since they are part of the tree.  I guess I need to add a specific version for that.
- if there are any problems with likelihood calculations, it will complain!
- mvSlice always moves, but it does more likelihood computations per move.

**Comment by bredelings** [comment]
So, I tried changing most of the mvScale and mvSlide moves to mvSlice, and it looks like there is a problem with `origin_time` where setting the origin time to its _current_ value makes the likelihood change by 15 log units.  So possibly there is a bug with likelihood calculation after all.

If I remove the slice moves on the origin time, then MCMC survives about 10 iterations, and then there is a problem updating `cont_clock_rate_mean`.  The likelihood is off by about 2 log units.

If I remove the slice moves on `cont_clock_rate_mean`, then I don't see any more problems with the MCMC.  I looked until generation 280 or so.

So, presumably there are some problems where changing `origin_time` and `cont_clock_rate_mean` is not updating or resetting the likelihood/prior properly.

**Comment by bredelings** [comment]
OK, so both problems have the same cause: the lnProbability for `cont_traits` is wrong.  Presumably something is failing to update it.

``` R
cont_traits ~ dnPhyloBrownianREML( ...)
```

So there is some kind of problem with the continuous part.  I'm not quite sure why DEBUG_MCMC didn't find it.  Presumably we now have to find which transition kernel is not updating the pdf for `dnPhyloBrownianREML( )`.

**Comment by bredelings** [comment]
This is *probably* fixed by #596, but we should probably check and make sure.

**Comment by davidcerny** [comment]
I've checked; it is indeed fixed. Here's the trace produced by `headers/MCMC/model_TED_w_backbone_const090624.Rev` on edeb6db:

<img width="1197" alt="Screenshot 2024-11-15 at 13 40 40" src="https://github.com/user-attachments/assets/1bf3ea87-49c1-4603-a770-ce8a0a7f81e1">

I've also tried:

1. Running a much longer analysis (~20.5 million moves) by swapping "MCMC_long" for "MCMC_test" in the script;
2. Re-running the short analysis (with "MCMC_test") under `--setOption debugMCMC=1`;
3. Re-running the long analysis under `--setOption debugMCMC=1` also;
4. Replacing the moves on `origin_time` and `cont_clock_rate_mean` with `mvSlice` in the short analysis and running it under `--setOption debugMCMC=1`.

No errors were encountered under debugging, and the parameter traces are not flatlining anymore. Closing this, with a huge thank you to @bredelings.
